### PR TITLE
Added afternoon/morning red tint

### PR DESCRIPTION
**Pull request by kaadmy**
I've added a reddish tint in the morning and afternoon; I know that it's not near perfect, but it's a start :)

### CODE REVIEW

File: src/mapblock_mesh.cpp [hunk 1]
```diff
@@ -335,12 +335,17 @@ u16 getSmoothLight(v3s16 p, v3s16 corner, MeshMakeData *data)
...
+	s32 b = r;
 
 	// Moonlight is blue
 	b += (day - night) / 13;
-	rg -= (day - night) / 23;
+	r -= (day - night) / 23;
+
+	// Morning/Afternoon is red
+	g -= MYMAX(0, 100 - abs(daynight_ratio - 500)) / 4;
+	b -= MYMAX(0, 100 - abs(daynight_ratio - 500)) / 4;
```
Comment (ShadowNinja) [inline src/mapblock_mesh.cpp]: `MYMAX(0, ...)` is equivalent to `abs(...)`, except that `MYMAX` evaluates its parameters twice.  Just use `abs`.

**Comment by ShadowNinja** [comment]
Sounds good.

File: src/mapblock_mesh.cpp (hunk 1 above)
Comment (est31) [inline src/mapblock_mesh.cpp]: `MYMAX(0,-2) == 0`. `abs(-2) = 2`.

Comment (ShadowNinja) [inline src/mapblock_mesh.cpp]: ... Of course.  I guess I wasn't thinking.

**Comment by paramat** [comment]
Nice idea i'll test. Can you describe when the red tint is added, how close to dawn/dusk it is?
I feel it should only be fairly close to dawn/dusk, during the light transitions, or, during the time the 'red glow' texture appears on the horizon, whichever is longer.

**Comment by kaadmy** [comment]
It appears when the time's around 5000-6000 and the sun's going up, around morning, and about the same time while the sun's going down.

**Comment by rubenwardy** [comment]
Here is a video of it in action: https://youtu.be/-zCvhgXe95w

**Comment by kaadmy** [comment]
Hmm, I can't see the red much; it's far more obvious for me.

Some images:
Early dawn: https://lut.im/t9A6p7DZS3/eO1ebx7TTTpPBuFu.png
Sunup: https://lut.im/IlnpzUAwdb/ozUa6G2WRUtOCKaf.png
Morning: https://lut.im/DpPi8i9zyk/JPjAEVDAQOU8Mvgs.png

**Comment by BlockMen** [comment]
:+1:  on the idea. Not tested though

**Comment by est31** [comment]
Like the idea as well.

**Comment by SmallJoker** [comment]
:+1: Looks much better now :)

**Comment by kaadmy** [comment]
Thanks for the compliments everybody :)
I can't believe that nobody else has thought of this before.

**Comment by kilbith** [comment]
Tested. Looks good :+1:

**Comment by BlockMen** [comment]
Hmm...without shaders the swap is kinda harsh, with shaders its not really noticable. Maybe it needs to be added there aswell? IIRC the boost of night-blue is done there aswell.

File: src/mapblock_mesh.cpp (hunk 1 above)
Comment (BlockMen) [inline src/mapblock_mesh.cpp]: You shouldn't pass an unsigned integer to abs() which breaks MSVC builds. Please convert ```daynight_ratio``` to ```s32``` first.

**Comment by kaadmy** [comment]
You can see the code that I've changed; the engine already tints the terrain at night, it just does not increase red during the morning/evening. This might increase CPU a little, but not swap.

**Comment by kaadmy** [comment]
Kinda bumping an oldish PR, but is this going to merged; and if not, why?

**Comment by rubenwardy** [comment]
This is such a small thing, I'd thought it had already been merged.

**Comment by est31** [comment]
@kaadmy can you do what @BlockMen  [said](https://github.com/minetest/minetest/pull/3299#discussion_r44225202)?

**Comment by paramat** [comment]
:-1: Something to consider: the redness will occur inside sealed structures (without windows) too and be a strange unwanted effect. Although blue is also boosted at night it is a slower-acting effect and is compensated for by the yellow-boost of artificial light.
I like the idea but unfortunately i have to change my opinion in this.

File: src/mapblock_mesh.cpp (hunk 1 above)
Comment (kaadmy) [inline src/mapblock_mesh.cpp]: Doesn't MT current pass `daynight_ratio` as a `u32` currently? I didn't change the type for that, I think.

**Comment by kaadmy** [comment]
Ah, hadn't though of the problem in structures, and underground aswell.

**Comment by est31** [comment]
@paramat isn't this something we separate light components for?

**Comment by kaadmy** [comment]
Hmm, I have another idea. Tint red only in places that are 15 light.

**Comment by paramat** [comment]
I don't think that would be worth the extra processing, also areas under trees and overhangs would be left untinted.
Any further idea on implementation or perhaps can this be closed?

**Comment by paramat** [comment]
Due to the issues i mention i don't think this can happen, any more opinions?

**Comment by Ferk** [comment]
@paramat, you said the blue at night is compensated by the yellow of the artificial lights.
I wonder if it would be possible to change the artificial light so that it also compensates the effect of dawn.

Maybe just store the G and B components increment caused by the time-of-day and substract it from the R B appliced to the artificial light in irr::core::clamp

Then the artificial light would always stay the same both in night and day.

File: src/mapblock_mesh.cpp (hunk 1 above)
Comment (HybridDog) [inline src/mapblock_mesh.cpp]: If you subtract 500 from daynight_ratio, the result is a signed int, else using the abs function here doesn't make sense, am l wrong?

**Comment by paramat** [comment]
Seems unjustifiable complexity to me.

File: src/mapblock_mesh.cpp (hunk 1 above)
Comment (HybridDog) [inline src/mapblock_mesh.cpp]: If you subtract 500 from daynight_ratio, the result, which is passed to abs, is a signed int, else using the abs function here doesn't make sense, am l wrong?

**Comment by Ferk** [comment]
It's literally 3 lines of code.

But I agree that the risk of having pink indoor lights is not worth the outdoors dawn effect. Also this would be out of control for subgames and maybe not every environment would want this effect.

**Comment by paramat** [comment]
Sure the code is simple, but it's applied to everything in meshes, so i think is best avoided unless really needed (which it isn't).

**Comment by benrob0329** [comment]
Tonemapping seems to tint the ground during sunrise/set.

**Comment by kaadmy** [comment]
Minetest already uses meshes to tint the ground at night blue, so I just changed a couple lines to add the red effect. I have no idea why people are still arguing why this is so inefficient, while MT already does this.

**Comment by paramat** [comment]
The PR is not inefficient, but it's in the bottleneck of meshgen so needs strong justification.
I think this should be in glsl nodes shader instead, there is already tinting code there already and at least then it's optional https://github.com/minetest/minetest/blob/master/client/shaders/nodes_shader/opengl_vertex.glsl
However there's still the problem of this being fast-acting so noticeable when acting inside sealed structures, that's going to be weird. The existing tints happen much slower at the speed of the day cycle so are not conciously noticeable.

**Comment by kaadmy** [comment]
The reason I added this in software is because my laptop is too slow to run with shaders, and with shaders on, fog is broken for some reason, likely having to do with an older/broken GLSL version.

**Comment by kaadmy** [comment]
_Bump_
Since I assume this is never going to be merged, should I close this PR?

**Comment by sfan5** [comment]
You could rebase it so it could be considered again

**Comment by juhdanad** [comment]
Once #5155 is implemented, closed structures with artificial light won't be tinted red.

**Comment by kaadmy** [comment]
Hm, that would help a lot.
This would probably have to be rewritten either way.

**Comment by paramat** [comment]
> Once #5155 is implemented, closed structures with artificial light won't be tinted red.

There's a high chance that won't happen. Even if it does i am still :-1:

**Comment by ShadowNinja** [comment]
Can't be implemented properly without #5155.
http://irc.minetest.net/minetest-dev/2017-04-15#i_4877053

**Comment by paramat** [comment]
Possibly replaced by commit by Ekdohibs:
https://github.com/Ekdohibs/minetest/commit/aedc81bf7b3999e9f166dad69d343821fbdce173

**Pull request closed without merging**